orttraining/orttraining/training_ops/cuda/communication/nccl_service.cu

```diff
@@ -0,0 +1,197 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#if defined(USE_NCCL) || defined(USE_HOROVOD)
+
+#include "orttraining/training_ops/cuda/communication/nccl_service.cuh"
+#include "core/profile/context.h"
+#include <iostream>
+#include <nccl.h>
+
+namespace onnxruntime {
+namespace cuda {
+
+void NcclService::SubmitSendAndWait(void* ptr, size_t size, int peer) {
+  // Wait until NCCL service is launched.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [this]{return is_launched_;});
+  }
+
+  auto& profile_context = profile::Context::GetInstance();
+  const auto tag = profile_context.GetThreadTagOrDefault(std::this_thread::get_id());
+  
+  // Pointer to enqueued task.
+  const NcclTask* task;
+
+  // Submit task.
+  {
+    std::lock_guard<std::mutex> guard(mutex_);
+    auto& profile_context = profile::Context::GetInstance();
+    const auto tag = profile_context.GetThreadTagOrDefault(std::this_thread::get_id());
+    task = schedule_[time_].EqueueTask(NcclTask::Type::SEND, std::vector<int>{peer}, ptr, size, tag);
+  }
+
+  // Wait for task to be finished.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [&]{return task->is_finished;});
+  }
+};
+
+void NcclService::SubmitRecvAndWait(void* ptr, size_t size, int peer) {
+  // Wait until NCCL service is launched.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [this]{return is_launched_;});
+  }
+
+  // Pointer to euqueued task.
+  const NcclTask* task;
+  {
+    std::lock_guard<std::mutex> guard(mutex_);
+    auto& profile_context = profile::Context::GetInstance();
+    const auto tag = profile_context.GetThreadTagOrDefault(std::this_thread::get_id());
+    task = schedule_[time_].EqueueTask(NcclTask::Type::RECV, std::vector<int>{peer}, ptr, size, tag);
+  }
+
+  // Wait for task to be finished.
+  {
+    std::unique_lock<std::mutex> lock(mutex_);
+    cv_.wait(lock, [&]{return task->is_finished;});
+  }
+};
+
+void NcclService::Initialize() {
+  // Here we assume GPU i is assigned to local process i.
+  // TODO: Create a general class to describe for computation topology and unify all similar uses.
+  // Hardwoards a process can own:
+  //   GPUs
+  //   CPUs
+  //   Other devices
+  int mpi_rank;
+  int mpi_size;
+  MPI_Comm_rank(MPI_COMM_WORLD, &mpi_rank);
+  MPI_Comm_size(MPI_COMM_WORLD, &mpi_size);
+
+  // Set device this NCCL communicator runs on.
+  cudaSetDevice(mpi_rank);
+
+  // Create stream to run NCCL.
+  cudaStreamCreate(&stream_);
+
+  // Get NCCL unique ID at rank 0 and broadcast it to all others.
+  ncclUniqueId id;
+  if (mpi_rank == 0) ncclGetUniqueId(&id);
+  MPI_Bcast((void *)&id, sizeof(id), MPI_BYTE, 0, MPI_COMM_WORLD);
+  ncclCommInitRank(&comm_, mpi_size, id, mpi_rank);
```

Comment:
Do we need error checking for MPI or CUDA call failures?

Reply:
Will add them. Thanks.

---
In reply to: [483800725](https://github.com/microsoft/onnxruntime/pull/5054#discussion_r483800725) [](ancestors = 483800725)